### DISCUSSION TITLE
Benefit of defining errors

### DISCUSSION THREAD

**Discussion by QzCurious** [ecyrbe/zodios#204] [Q&A] [answered]
Apart from validating schema and throw ZodiosError. I'm wondering how can I do it better with zodios by defining errors.
I'm assume that because zodios already parse error response for me. I should be able to narrowing down the error when it thrown from zodios request.

Below is what I usually do. Is zodios capable of making `const res = err.response.data as LoginError` more type safe?
```ts
interface LoginError {
  code: "not found" | "locked";
}

async function login() {
  try {
    await axios.post("/login", { account: "qzcurious", password: "xxxxxxxx" });
  } catch (err) {
    if (!axios.isAxiosError(err)) throw err;
    // `err.response` is now type AxiosResponse

    if(!err.response) throw err

    const res = err.response.data as LoginError; // <-- I want improve this
    if (res.code === "locked") {
        toast('user is locked, please contact with your supervisor')
    }
    else if(res.code === 'not found') {
        toast('User not found.')
    }
  }
}
```

**Comment by ecyrbe** [comment]
Zodios does not parse errors today. 
The rationale being that this would just make a New error. And also that errors can be unexpected, their format can also be, so parsing is likely to fail on some unexpected scenarios.

So Error handling is inhently not type safe. 
I haven't yet found the solution to this problem. All ideas are welcomed.
For the moment i'm thinking about a helper that does matching instead of just parsing, like this:
```ts
const { error, status, type } = matchZodiosError(e);
switch (type) {
  case 'ValidationError':  // type will be inferred
  case 'ExpectedRequestError':   // type will be inferred
  case 'UnexpectedRequestError' : // type will not be inferred
  default : // not an error from zodios
}
```

**Reply by QzCurious** [reply]
Could I ask you some code example for what would you do for handling the error response? Do you also doing the type assertion?

**Reply by ecyrbe** [reply]
Here is a pull request of this idea :  https://github.com/ecyrbe/zodios/pull/206

**Comment by QzCurious** [comment]
For my point of view, it is two individual topic, one is **parsing**, another is **type narrowing**. Let me talk about parsing first:

I think we should parse error response too:

For example
1. POST /login
2. It could response with 404 with body `{error: 'USER_NOT_FOUND'}`
3. The schema is known on author time, not runtime.
4. So we should `makeErrors` for it

And how should error be handled:

a. If the error fail to be parsed
    `ZodiosError` should be throw, because it stands for "parsing failed" or "untursed response format"
b. If the error is parsed successfully
    Just re-throw what axios throws (or something inherit AxiosError that can also help with type narrowing)

_I'm trying to describe it more correctly. If there any wording seem to be offensive, I'm sorry, and please tell me._

**Reply by ecyrbe** [reply]
i have added an example in the PR. you can give feedback there if you want : https://github.com/ecyrbe/zodios/pull/206

**Comment by QzCurious** [comment, marked as answer]
🎉 The answer for the Q&A is implemented on a new [error handling feature](https://www.zodios.org/docs/client/error).

**Reply by QzCurious** [reply]
The moment the autocompletion prompt by
```ts
if (isErrorFromAlias('login', err)) {
  if (err.response.data.code === '') {}
                              // ^^ it's autocompleting!!!
}
```
is like a dream come true.
What more excited is that **it's not a dream, it is happening!!!**
